core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -753,7 +753,7 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
    * Keep or break a promise based on the result of this action.
    */
   final def to[E1 >: E, A1 >: A](p: Promise[E1, A1]): IO[Nothing, Boolean] =
-    self.run.flatMap(x => p.done(IO.done(x)))
+    self.redeem(p.error, p.complete).onTermination(cause => p.done(IO.fail0(cause)))
```

Comment:
This should work great (although in some cases the promise will already be set when the `onTermination` handler is called—that's fine, `done` will just return false), but two things to keep in mind:

1. `onTermination` is built on the most powerful `bracket0` which forks and does all that other stuff, it will probably be quite slow, and certainly slower than `onInterrupt`
2. I am really interested in changing the semantics of `run` so that if the `io` is interrupted, you see that reflected in the `ExitResult`

I know @kaishh expressed reservations about (2). However, as we discussed at dinner, the `IO` returned from `run` can still be interrupted. It's just that if the inner `IO` joined something that was interrupted (e.g. an interrupted promise), then that interruption will be captured by the `ExitResult` returned from `run`.

It's a distinction between "inner interruption" and "outer interruption". The IO returned from `run` can still be interrupted from the outside, and this in turn should interrupt the inner `IO`. I think we can achieve that without too much work and satisfy @kaishh's requirements but also make `run` a lot easier to understand / use.

Reply:
Yep, let me try that 👍